### DISCUSSION TITLE
Rust migration

### DISCUSSION THREAD

**Discussion by deeprobin** [dotnet/runtime#49602] [Ideas] [answered]
### Description

I would have the idea to improve the CLR performance technically still something. Reference Counting is no longer 2021, why don't we use an ownership system for the CLR.

### Configuration

General issue / not important

### Regression?

I think not.

### Data

* [Introduction to Rust](https://serokell.io/blog/rust-guide)
* [Rust vs C++ comparison](https://www.apriorit.com/dev-blog/520-rust-vs-c-comparison)
* [Best practices / Cheat Sheet for Rust](https://cheats.rs/)
* [Benchmarks: Rust vs C++](https://benchmarksgame-team.pages.debian.net/benchmarksgame/fastest/rust.html)

### Analysis

You could migrate the project to Rust. You can also migrate only subprojects that make a lot of difference in terms of performance. Rust has builtin FFI support.

**Comment by ghost** [comment]
I couldn't figure out the best area label to add to this issue. If you have write-permissions please help me learn by adding exactly one [area label](https://github.com/dotnet/runtime/blob/master/docs/area-owners.md).

**Comment by huoyaoyuan** [comment]
You may be surprised that **CLR hasn't fully migrated to C++**. It was started written before 1999, even before the first C++ standard. There are still many C-style snippets.

> Reference Counting is no longer 2021, why don't we use an ownership system for the CLR.

Ownership isn't perfect. It can't handle long-lived objects well. I'm not seeing CLR bothered by object lifetime.

In my sights, Rust isn't suitable for CLR. CLR is manipulating the exact memory layout of objects and registers. And it must be unsafe across managed-unmanaged boundary.

**Reply by jkotas** [reply]
And we are migrating the C/C++ code to C# as necessary. It avoids overhead of crossing managed-unmanaged boundary and often reduces use of unsafe.

**Comment by ufcpp** [comment, marked as answer]
https://github.com/dotnet/csharplang/issues/3860#issuecomment-688998222
